### PR TITLE
Detect culture change in .NET

### PR DESCRIPTION
**Pull request by pranavkm**
With ICU sharding enabled, blazor wasm attempts to detect if the application culture was changed by the
application code as part of Program.MainAsync and tell them they need to opt out of sharding.

Prior to this change, Blazor compared the .NET culture string with a JS representation for language. With iOS 14,
the two culture strings differ in casing which prevents the use of any Blazor WASM app the latest version installed.

As part of this change, the comparison is performed entirely in .NET which avoids relying on the JS representation.

---------------------------------

### Description

Blazor WebAssembly has a check to verify if the application CultureInfo was modified as part of Program.Main. This check produces false positives on iOS 14 and prevents the app from starting.

### Customer impact
Unless developers modify their project to workaround the false positive, apps no longer start on the newest iOS.

### Regression

Yes. This was a result of ICU data sharding introduced in rc2 and a recent iOS update.

### Risk

Low. The changes are isolated to WASM's bootstrapping code and are made more robust because it's now entirely in .NET (versus being a mix of .NET and JS). We have good unit and functional test coverage to verify that the false positives are removed and manually verified that Blazor apps load on iOS 14.

### CODE REVIEW

File: src/Components/Web.JS/src/Boot.WebAssembly.ts [hunk 1]
```diff
@@ -8,7 +8,7 @@ import { shouldAutoStart } from './BootCommon';
 import { setEventDispatcher } from './Rendering/RendererEventDispatcher';
 import { WebAssemblyResourceLoader } from './Platform/WebAssemblyResourceLoader';
 import { WebAssemblyConfigLoader } from './Platform/WebAssemblyConfigLoader';
-import { BootConfigResult } from './Platform/BootConfig';
+import { BootConfigResult, ICUDataMode } from './Platform/BootConfig';
```
Comment (captainsafia) [inline src/Components/Web.JS/src/Boot.WebAssembly.ts:11]: Nit: Don't think this is needed here.

File: src/Components/WebAssembly/WebAssembly/src/Hosting/EntrypointInvoker.cs [hunk 2]
```diff
@@ -10,13 +11,17 @@ namespace Microsoft.AspNetCore.Components.WebAssembly.Hosting
 {
     internal static class EntrypointInvoker
     {
+        private const string NotifyCultureChanged = "window.Blazor._internal.notifyCultureChanged";
```
Comment (captainsafia) [inline src/Components/WebAssembly/WebAssembly/src/Hosting/EntrypointInvoker.cs:14]: Is this actually used anywhere?

Comment (campersau) [inline src/Components/WebAssembly/WebAssembly/src/Hosting/EntrypointInvoker.cs:14]: Is this used anywhere?

**Review by captainsafia — APPROVED** [review]
Seems good overall. Do we need any more modifications to the tests?

File: src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts [hunk 3]
```diff
@@ -404,6 +399,14 @@ function createEmscriptenModuleInstance(resourceLoader: WebAssemblyResourceLoade
     }
     resourceLoader.purgeUnusedCacheEntriesAsync(); // Don't await - it's fine to run in background
 
+    if (resourceLoader.bootConfig.icuDataMode == ICUDataMode.Sharded) {
```
Comment (SteveSandersonMS) [inline src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:402]: Nit:

```suggestion
    if (resourceLoader.bootConfig.icuDataMode === ICUDataMode.Sharded) {
```

File: src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts [hunk 4]
```diff
@@ -404,6 +399,14 @@ function createEmscriptenModuleInstance(resourceLoader: WebAssemblyResourceLoade
...
+      MONO.mono_wasm_setenv('__BLAZOR_SHARDED_ICU',  '1');
+
+      if (resourceLoader.startOptions.applicationCulture) {
+        // If a culture is specified via start options use that to initialize the Emscripten \  .NET culture.
+        MONO.mono_wasm_setenv('LANG',  `${resourceLoader.startOptions.applicationCulture}.UTF-8`);
```
Comment (SteveSandersonMS) [inline src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:407]: This is interesting. Were we missing a feature before? How do you know this magic string format is correct - where does the info come from?

**Review by SteveSandersonMS — APPROVED** [review]
Nice!

File: src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts (hunk 4 above)
Comment (pranavkm) [reply src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:407]: This was one of the fallouts of the change. Previously this value influenced the icuData string and we expected the user to configure the .NET culture again using the same mechanism they used to determine the culture in JS. The Globalalization test app used this. It seemed slightly more preferable to constructing magic strings. 

However this trips our current detection so it seems worthwhile to initialize .NET with the right culture all the way. The string format is based on emscripten's implementation: https://github.com/emscripten-core/emscripten/blob/d9c79342262a0b05175ad1c029f4528bc8e1d136/src/library_wasi.js#L20

**Pull request merged**